src/types/array.rs

```diff
@@ -14,6 +14,11 @@
 //! |   1   | `i32`          | Array end offset[^2].    |
 //! |   2   | `i32`          | Allocated capacity[^2].  |
 //!
+//! The pointer to the allocation (which is **not the data**) contains:
+//!   1. Reference counter.
+//!   2. Padding.
+//!   3. Array data. Its address is the pointer to the data stored in the type.
+//!
```

Comment:
I'm not sure I understood it correctly. What do you mean by _allocation_? 

If the index 0 points to the array data, then to calculate the counter, you need to substract (padding + counter size) from the data pointer?

Reply:
The call to `malloc` returns a pointer, which is the allocation pointer; aka, the pointer pointing to the allocation. This pointer is not what is stored within the array struct.

The allocation pointer has:
  - The reference counter (i32).
  - Padding to align the data.
  - The actual data (the pointer to which is stored in the array struct).

Previously the allocation pointer was the same as the data pointer, but this has been changed. Now, to free the array we need to subtract the offset calculated by the reference counter and the padding. This has been made because most of the time we don't need the reference counter and having to offset the pointer every time would add boilerplate and unnecessary operations.

> If the index 0 points to the array data, then to calculate the counter, you need to substract (padding + counter size) from the data pointer?

Yes.